File: src/helpers/images.py
import io
from PIL import Image
import math

import requests

def _resize_on_height(img: Image, max_height: int) -> Image:
  ratio = max_height / img.height
  new_width = math.floor(img.width * ratio)

  return img.resize((new_width, max_height))

def _resize_on_width(img: Image, max_width: int) -> Image:
  ratio = max_width / img.width
  new_height = math.floor(img.height * ratio)

  return img.resize((max_width, new_height))

def fit_image_to_constraint(image: Image, constraint=4096) -> Image:
  if (image.height > constraint and image.height > image.width):
    return _resize_on_height(image, constraint)
  elif (image.width > constraint):
    return _resize_on_width(image, constraint)
  else:
    # do nothing
    return image

def get_image(url: str) -> Image:
  headers = {'User-Agent': 'Wikimedia Bot' }
  resp = requests.get(url, headers=headers)

  print(f'downloading {url}')

  if resp.status_code != 200:
      raise Exception('Something went wrong downloading the file!')
  return Image.open(io.BytesIO(resp.content))

def to_bytes(image: Image) -> bytes:
  format = image.format or 'JPEG'
  image_data = io.BytesIO()
  print(format)
  image.save(image_data, format=format)
  return image_data.getvalue()

File: src/helpers/images.ts
import sharp from 'sharp';
import axios from 'axios';

const resizeOnHeight = async (
  img: sharp.Sharp,
  maxHeight: number
): Promise<sharp.Sharp> => {
  const metadata = await img.metadata();
  if (!metadata.height || !metadata.width)
    throw new Error('invalid image metadata');

  const ratio = maxHeight / metadata.height;
  const newWidth = Math.floor(metadata.width * ratio);

  return img.resize(newWidth, maxHeight);
};

const resizeOnWidth = async (
  img: sharp.Sharp,
  maxWidth: number
): Promise<sharp.Sharp> => {
  const metadata = await img.metadata();
  if (!metadata.height || !metadata.width)
    throw new Error('invalid image metadata');

  const ratio = maxWidth / metadata.width;
  const newHeight = Math.floor(metadata.height * ratio);

  return img.resize(maxWidth, newHeight);
};

export const fitImageToConstraint = async (
  image: sharp.Sharp,
  constraint = 4096
): Promise<sharp.Sharp> => {
  const metadata = await image.metadata();
  if (!metadata.height || !metadata.width)
    throw new Error('invalid image metadata');

  if (metadata.height > constraint && metadata.height > metadata.width) {
    return resizeOnHeight(image, constraint);
  } else if (metadata.width > constraint) {
    return resizeOnWidth(image, constraint);
  }
  return image;
};

export const getImage = async (url: string): Promise<sharp.Sharp> => {
  const headers = { 'User-Agent': 'Wikimedia Bot' };
  const response = await axios.get(url, {
    headers,
    responseType: 'arraybuffer',
  });

  console.log(`downloading ${url}`);

  if (response.status !== 200) {
    throw new Error('something went wrong downloading the file!');
  }

  return sharp(response.data);
};

export const toBlob = async (image: sharp.Sharp): Promise<Blob> => {
  const metadata = await image.metadata();
  const format = metadata.format || 'jpeg';
  const buffer = await image.toBuffer();
  return new Blob([buffer], { type: `image/${format}` });
};

export const toBuffer = async (image: sharp.Sharp): Promise<Buffer> => {
  const metadata = await image.metadata();
  const format = metadata.format || 'jpeg';
  console.log(format);
  return image.toBuffer();
};
